File: backend/api/schemas/inference.py
from pydantic import BaseModel
from typing import Dict, List, Optional

class EstimatorOptions(BaseModel):
    mean_estimators: List[str]
    deviation_estimators: List[str]

class EstimatorRequest(BaseModel):
    session_id: str
    column: str

class InferenceParams(BaseModel):
    session_id: str
    column: str
    alpha: float = 0.05
    mean_estimator: str = "Sample Mean"
    median_estimator: str = "Sample Median"
    sigma_estimator: str = "Deviation (1 ddof)"
    trim_param: Optional[float] = None
    winsor_limits: Optional[str] = None
    weights_column: Optional[str] = None
    bootstrap_mean: bool = False
    bootstrap_median: bool = False
    bootstrap_deviation: bool = False
    bootstrap_pi: bool = False
    bootstrap_samples: int = 1000
    filters: Optional[Dict[str, List[str]]] = None

class ConfidenceRegionsParams(InferenceParams):
    probs: List[float]
    eps_mu: List[float]
    eps_sigma: List[float]
    add_ci_box: bool = True
    mu_ci_source: str = "Mean-based CI"

class IntervalsResponse(BaseModel):
    table: str # JSON string of DataFrame via to_json(orient="records")

class ConfidenceRegionsResponse(BaseModel):
    z_matrix: List[List[float]]
    mu_grid: List[float]
    sigma_grid: List[float]
    mu_hat: float
    sigma_hat: float
    mean_ci: List[float]
    sigma_ci: List[float]
    probs: List[float]
    levels: List[float]
    table: Optional[str] = None # Optional tabular view when add_ci_box is True

File: frontend/src/api/inference.ts
const BASE_URL = "/api/inference";

export interface EstimatorRequest {
  session_id: string;
  column: string;
}

export interface EstimatorOptions {
  mean_estimators: string[];
  deviation_estimators: string[];
}

export interface InferenceParams {
  session_id: string;
  column: string;
  alpha: number;
  mean_estimator: string;
  median_estimator: string;
  sigma_estimator: string;
  trim_param?: number | null;
  winsor_limits?: string | null;
  weights_column?: string | null;
  bootstrap_mean: boolean;
  bootstrap_median: boolean;
  bootstrap_deviation: boolean;
  bootstrap_pi: boolean;
  bootstrap_samples: number;
  filters?: Record<string, string[]> | null;
}

export interface ConfidenceRegionsParams extends InferenceParams {
  probs: number[];
  eps_mu: number[];
  eps_sigma: number[];
  add_ci_box: boolean;
  mu_ci_source: string;
}

export interface IntervalsResponse {
  table: string; // JSON string
}

export interface ConfidenceRegionsResponse {
  z_matrix: number[][];
  mu_grid: number[];
  sigma_grid: number[];
  mu_hat: number;
  sigma_hat: number;
  mean_ci: number[];
  sigma_ci: number[];
  probs: number[];
  levels: number[];
  table?: string;
}

export const inferenceApi = {
  getEstimators: async (req: EstimatorRequest): Promise<EstimatorOptions> => {
    const response = await fetch(`${BASE_URL}/estimators`, {
      method: "POST",
      headers: { 
        "Content-Type": "application/json",
        "x-session-id": req.session_id
      },
      body: JSON.stringify(req),
    });
    if (!response.ok) {
      const text = await response.text();
      let errStr = text;
      try { errStr = JSON.parse(text).detail || text; } catch {}
      throw new Error(errStr);
    }
    return response.json();
  },

  computeCI: async (params: InferenceParams): Promise<IntervalsResponse> => {
    const response = await fetch(`${BASE_URL}/ci`, {
      method: "POST",
      headers: { 
        "Content-Type": "application/json",
        "x-session-id": params.session_id
      },
      body: JSON.stringify(params),
    });
    if (!response.ok) {
      const text = await response.text();
      let errStr = text;
      try { errStr = JSON.parse(text).detail || text; } catch {}
      throw new Error(errStr);
    }
    return response.json();
  },

  computePI: async (params: InferenceParams): Promise<IntervalsResponse> => {
    const response = await fetch(`${BASE_URL}/pi`, {
      method: "POST",
      headers: { 
        "Content-Type": "application/json",
        "x-session-id": params.session_id
      },
      body: JSON.stringify(params),
    });
    if (!response.ok) {
      const text = await response.text();
      let errStr = text;
      try { errStr = JSON.parse(text).detail || text; } catch {}
      throw new Error(errStr);
    }
    return response.json();
  },

  computeRegions: async (params: ConfidenceRegionsParams): Promise<ConfidenceRegionsResponse> => {
    const response = await fetch(`${BASE_URL}/regions`, {
      method: "POST",
      headers: { 
        "Content-Type": "application/json",
        "x-session-id": params.session_id
      },
      body: JSON.stringify(params),
    });
    if (!response.ok) {
      const text = await response.text();
      let errStr = text;
      try { errStr = JSON.parse(text).detail || text; } catch {}
      throw new Error(errStr);
    }
    return response.json();
  },
};
